cpp/src/cider/type/plan/StringOpExpr.cpp

```diff
@@ -316,54 +316,106 @@ JITExprValue& SubstringStringOper::codegen(CodegenContext& context) {
-            .ret_type = JITTypeTag::POINTER,
-            .params_vector = {string_heap_ptr.get(), ret_len.get()}});
-    auto ret_ptr = func.emitRuntimeFunctionCall(
-        fn_name + "_ptr",
-        JITFunctionEmitDescriptor{
-            .ret_type = JITTypeTag::VOID,
-            .params_vector = {
-                ptr.get(), arg_val.getValue().get(), pos_param.get(), len_param.get()}});
-    return set_expr_value(arg_val.getNull(), ret_len, ptr);
+    auto ret_ptr_int64 = func.createVariable(JITTypeTag::INT64);
+    auto ret_len = func.createVariable(JITTypeTag::INT32);
+    if_builder
+        ->condition([&]() {
+          auto condition = arg_val.getNull();
+          return condition;
+        })
+        ->ifFalse([&]() {
+          // format parameters
+          auto pos_param = func.emitRuntimeFunctionCall(
+              "format_substring_pos",
+              JITFunctionEmitDescriptor{
+                  .ret_type = JITTypeTag::INT32,
+                  .params_vector = {pos_val.getValue()
+                                        ->castJITValuePrimitiveType(JITTypeTag::INT32)
+                                        .get(),
+                                    arg_val.getLength().get()}});
+          auto len_param = func.emitRuntimeFunctionCall(
+              "format_substring_len",
+              JITFunctionEmitDescriptor{
+                  .ret_type = JITTypeTag::INT32,
+                  .params_vector = {pos_param.get(),
+                                    arg_val.getLength().get(),
+                                    len_val.getValue()
+                                        ->castJITValuePrimitiveType(JITTypeTag::INT32)
+                                        .get()}});
+
+          // get string heap ptr
+          auto string_heap_ptr = func.emitRuntimeFunctionCall(
+              "get_query_context_string_heap_ptr",
+              JITFunctionEmitDescriptor{.ret_type = JITTypeTag::POINTER,
+                                        .ret_sub_type = JITTypeTag::INT8,
+                                        .params_vector = {func.getArgument(0).get()}});
+          *ret_len = *len_param;
+
+          auto ret_ptr = func.emitRuntimeFunctionCall(
+              fn_name + "_ptr",
+              JITFunctionEmitDescriptor{
+                  .ret_type = JITTypeTag::POINTER,
+                  .params_vector = {arg_val.getValue().get(), pos_param.get()}});
+          *ret_ptr_int64 = *func.emitRuntimeFunctionCall(
+              "cast_ptr_to_int64",
+              JITFunctionEmitDescriptor{.ret_type = JITTypeTag::INT64,
+                                        .params_vector = {ret_ptr.get()}});
+        })
+        ->build();
+    return set_expr_value(
+        arg_val.getNull(),
+        ret_len,
+        func.emitRuntimeFunctionCall(
+            "cast_int64_to_ptr",
+            JITFunctionEmitDescriptor{.ret_type = JITTypeTag::POINTER,
+                                      .params_vector = {ret_ptr_int64.get()}}));
```

Comment:
Seems several `if...else...` branches are added, will it cause perf degradation if null_ratio is small and prevent future vectorization?

Reply:
`if (isOutput()) {} else ` will generate different code depending on expression type, this will not cause any perf degradation. `ifBuilder` will generating two paths for null and not-null rows. Actually, it will short circuit for null path, performance is better as far as I test.  This may influence primitive type vectorization but will not for string type.

Reply:
I didn't do such test. but it did better for ratio > 0 case.